### DISCUSSION TITLE
How can i store all the LUT and leaf cells used to implement a design in to a text file ?

### DISCUSSION THREAD

**Discussion by SSandeep19** [Xilinx/RapidWright#456] [Q&A]
I have a design check point file post implementation in Vivado holding the LUT information. I want to save theat information in to text file. The text file should have information like the name of the LUT the slice of the FPGA etc. Is it possible to do using rapid write ?. 

Thank you.

**Comment by clavin-xlnx** [comment]
Sure, here is some example code that loads in one of our test DCPs and write outs a text file:

```java
package com.xilinx.rapidwright.examples;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;

import com.xilinx.rapidwright.design.Cell;
import com.xilinx.rapidwright.design.Design;

public class WriteLUTInfoToFile {
    public static void main(String[] args) throws IOException {
        Design d = Design.readCheckpoint("test/RapidWrightDCP/bnn.dcp");
        try(BufferedWriter bw = new BufferedWriter(new FileWriter("lutInfo.txt"))){
            for(Cell c : d.getCells()) {
                if(!c.getBEL().isLUT()) continue;
                bw.write(c.getName() + " " + c.getSiteName() + "/" + c.getBELName() + " " + c.getType() + "\n");
            }            
        }
    }
}

```

**Comment by SSandeep19** [comment]
Thank you for the reply. I am new to this environment and little confused i hope you can help me out here. I did automatic install setup and got the IDE as shown in below screen shot.

My goal is to take a post implementation design check point file from Vivado and get all the slices used by the design into a text file. The FPGA i am using is kc705 (xc7k325ffg900).

What should i do after installation ? I see there are couple of IDEs do i need them but i already have Vivado. Is there any example you can provide which is similar to my task.

Any help is much appreciated. Thank you

![image](https://user-images.githubusercontent.com/56548200/172873416-23f21cb9-0f99-4548-90ba-1d252a4da392.png)

**Reply by clavin-xlnx** [reply]
RapidWright is more of a library to build upon than an actual application.  The GUI that comes up as part of the install is simply a tool to example devices present in RapidWright.  I would recommend that you take the example I posted previously and save it to a `.java` file in your RapidWright install (for example: `RapidWright/src/com/xilinx/rapidwright/examples/WriteLUTInfoToFile.java`).  Then you can compile and run it (either via IDE or command line).

**Reply by SSandeep19** [reply]
I am facing two problems

One - There is no bnn.dcp file so i have uploaded my own post implementation vivado check point file

Two - When i compile the java file i am facing errors as shown in below screen shot. (WriteLUTInfoToFile.java:7: error: package com.xilinx.rapidwright.design does not exist import com.xilinx.rapidwright.design.Cell;) etc. Received similar kind of error post manual installation but able to open the IDE shown above.

![image](https://user-images.githubusercontent.com/56548200/173191757-aa897f52-e5c4-4134-a421-15b7ada706cc.png)

**Reply by SSandeep19** [reply]
Below is the error what i got if i used IDE IntelliJ

![image](https://user-images.githubusercontent.com/56548200/173194736-d46e17c5-ad1b-41b0-bd7c-94fbacb1f9fa.png)

**Reply by clavin-xlnx** [reply]
> I am facing two problems
> 
> One - There is no bnn.dcp file so i have uploaded my own post implementation vivado check point file
> 
The `bnn.dcp` file is not present probably because the git submodule hasn't been populated.  You can do this by running `git submodule update --init` or using our built-in gradle task `./gradlew updateSubmodules`.  By default, `git` doesn't automatically pull submodules in a repo.  


> Two - When i compile the java file i am facing errors as shown in below screen shot. (WriteLUTInfoToFile.java:7: error: package com.xilinx.rapidwright.design does not exist import com.xilinx.rapidwright.design.Cell;) etc. Received similar kind of error post manual installation but able to open the IDE shown above.
> 

The compile errors indicate that the `rapidwright-api-lib.jar` is not on the compiler's `CLASSPATH`.  An easy way to ensure the code is compiled correctly from the command line is to use gradle `./gradlew compileJava`.  

Not sure what is going on in IntelliJ - @eddieh-xlnx any thoughts?

**Reply by SSandeep19** [reply]
Thank you for the response.
First problem is solved.
For second one i couldnt find rapidwright-api-lib.jar, where can i get ti. I tried gradlew compilejava and below is the error.
![image](https://user-images.githubusercontent.com/56548200/173426077-cea912fe-12cc-4741-b37d-cd967312075d.png)
I tried it on the rapidwright folder below is the output of the terminal.
![image](https://user-images.githubusercontent.com/56548200/173426267-8c2b3dc6-1787-4b4d-bd5c-cfc0b567a1f0.png)

**Reply by clavin-xlnx** [reply]
You need to be in the project directory, in your case `C:\rw2\RapidWright` for any Gradle task to run.  Also note that you want to run `gradlew compileJava` (note the capital 'J').

**Reply by clavin-xlnx** [reply]
If you run `gradlew compileJava`, you don't need to invoke `javac`, it takes care of that for you.  If you want to run the example, run:
```
java -cp bin;jars\* com.xilinx.rapidwright.examples.WriteLUTInfoToFile
```
This adds the compiled source code and jar files to the CLASSPATH so `java` knows where to find the class files involved in running the example.

**Reply by SSandeep19** [reply]
But the Java file is in the directory - RapidWright/src/com/xilinx/rapidwright/examples/WriteLUTInfoToFile.java. I took the java file and copied into RapidWright I am facing the error given below.
![image](https://user-images.githubusercontent.com/56548200/173429853-6e42ce62-4907-4789-b8e7-9afa5984a442.png)
I pasted file in the rapidwright directory but still the error exists.

**Reply by clavin-xlnx** [reply]
It looks like you are running the command `gradlew compileJava WriteLUTInfoToFile.java`, this is incorrect.  You want to compile all the code by running `gradlew compileJava`.  Give that a try and then run the `java` command I mentioned in my past comment.

**Reply by SSandeep19** [reply]
Thank you for your patience and response. 
I am able to execute the program but where can i find the LUT information below is how the terminal looks like.
![image](https://user-images.githubusercontent.com/56548200/173433035-b57623a0-006e-4ff3-947c-c5e2f17c0e90.png)

**Reply by SSandeep19** [reply]
Found it thanks.

**Reply by SSandeep19** [reply]
To get all the flip flops information can i just change 
if(!c.getBEL().isLUT()) continue;                      to                 if(!c.getBEL().isFDCE()) continue;

Will that work ?

**Reply by clavin-xlnx** [reply]
`BEL.isFDCE()` is not a method that I am aware of.  However, there is [`BEL.isFF()`](http://www.rapidwright.io/javadoc/com/xilinx/rapidwright/device/BEL.html#isFF--).

**Reply by SSandeep19** [reply]
Got it thanks!! Is there any way to write all the elements used by the design ? (LUT,FF,BRAM,DSP etc)
